frontend/src/metabase/lib/engine.ts

```diff
@@ -0,0 +1,68 @@
+import type { Engine } from "metabase-types/api";
+
+export function getEngineNativeType(engine?: string) {
+  switch (engine) {
+    case "mongo":
+    case "druid":
+      return "json";
+    default:
+      return "sql";
+  }
+}
+
+export function getNativeQueryLanguage(engine?: string) {
+  return getEngineNativeType(engine).toUpperCase();
+}
+
+export function getEngineNativeAceMode(engine?: string | null) {
+  switch (engine) {
+    case "mongo":
+    case "druid":
+      return "ace/mode/json";
+    default:
+      return "ace/mode/sql";
+  }
+}
+
+type JSONQuery = Record<string, any> | Record<string, any>[];
+
+function formatJsonQuery(query: JSONQuery) {
+  return JSON.stringify(query, null, 2);
+}
+
+export function formatNativeQuery(query?: string | JSONQuery, engine?: string) {
+  if (!query || !engine) {
+    return;
+  }
+
+  if (getEngineNativeType(engine) === "sql" && typeof query === "string") {
+    return formatSQL(query);
+  } else if (
+    getEngineNativeType(engine) === "json" &&
+    typeof query === "object"
+  ) {
+    return formatJsonQuery(query);
+  }
```

Comment:
Not all paths explicitly return value from this function.

I think `return query;` would be a safe callback.
We could also use it in the first `if` statement of this function.

Reply:
I can't return the `query` because the functions that depend on the returned result from this function require a string.

I can either return an empty string or `Throw`.

Reply:
> I can't return the query because the functions that depend on the returned result from this function require a string.

Line 35 returns undefined, not a string. If it works, we can do it here too (it's already happening implicitly but let's be explicit about it).